doc/contribute/index.rst

```diff
@@ -345,6 +345,43 @@ following exceptions:
   comment, ``//``, is not allowed.
 * Use ``/**  */`` for doxygen comments that need to appear in the documentation.
 
+The C standard reserves all identifiers that begin with an underscore
```

Comment:
I'm not sure we need all the C standard backup explanation (TLDR), so how about replacing this whole section with something like this (we'll need some developer review that I've got this right):
```
Identifier Naming
=================

* Names of macros defining constants and labels in enums
  are all capital letters and underscores.
* Functions,  and function-like macros use all lowercase,
  with underscores, of a form "module_verb_noun", for example
  ``counter_get_frequency``, and ``coap_make_response_code``
* All other identifiers including variables and types use all 
   lowercase and underscores.
* Do not use mixed-case CamelCaseNaming.

leading underscore for low level functions
The C standard (7.1.3 [ISO/IEC 9899:2011]) reserves all identifiers
that begin with an underscore for implementation details.  As such,
identifiers in Zephyr must not start with an underscore.
Instead, we address the traditional Linux kernel use of leading
underscore for internal names or
low-level functions by using a leading ``z_``, for example:

* ``_arch_`` becomes ``z_arch_``
* ``_sys_`` becomes  ``z_sys_``
* ``_k_`` becomes  ``z_``
* ``_impl_`` becomes ``z_impl_``
* ``_handler_`` becomes ``z_handl\_``
* and other leading underscore cases become ``z_``

Coding style enforcement
========================
```

(Add this last heading to lead into the checkpatch discussion)

Reply:
Most function-like macros seem to use uppercase. Can do a `git grep '#define.*[a-zA-Z]('` to see a bunch of them.

Looks alright to me otherwise. Maybe "identifiers in Zephyr" could be replaced with "global identifiers in Zephyr", to avoid some confusion without going full standardese.

Think I saw either @andyross or @andrewboie write something about naming recently.

Reply:
Hmm... wondering if `z_`/`Z_` is really meant for internal names though. I thought it was for public APIs.

If we can't have `__` and stuff for internal functions, maybe they could have an `_imp` suffix instead. Seems confusing to put `z_` before them at least, if that marks an API.

Reply:
I'll update the uppercase. About prefix _imp as suffix works, the only thing is that we had already a patch addressing it (using z_impl). What was discussed is that z_impl is the namespace, same for z_arch, z_sys, ...

Not opposing the change, it should be easy change it. I'm just adding this to reflect what is happening now

Reply:
> wondering if z_/Z_ is really meant for internal names though. I thought it was for public APIs.

z_/Z_ is for private APIs exposed in headers which should not be relied on by an application because they could disappear or change at any time.

I don't think we need a naming convention for static functions at all, other than to not violate the C standard.